cleverhans/attacks_eager.py

```diff
@@ -0,0 +1,260 @@
+        return labels, nb_classes
+
+    def generate_np(self, x_val, **kwargs):
+        """
+        Generate adversarial examples and return them as a NumPy array.
+
+        :param x_val: A NumPy array with the original inputs.
+        :param **kwargs: optional parameters used by child classes.
+        :return: A NumPy array holding the adversarial examples.
+        """
+        tfe = tf.contrib.eager
+        x = tfe.Variable(x_val)
+        adv_x = self.generate(x, **kwargs)
+        return adv_x.numpy()
+
+    def clip_eta(self, eta, ord, eps):
+        """
+        Helper function to clip the perturbation to epsilon norm ball.
+        :param eta: A tf variable with the current perturbation.
+        :param ord: Order of the norm (mimics Numpy).
+                    Possible values: np.inf, 1 or 2.
+        :param eps: Epilson, bound of the perturbation.
+        """
+        # Clipping perturbation eta to self.ord norm ball
+        if ord not in [np.inf, 1, 2]:
+            raise ValueError('ord must be np.inf, 1, or 2.')
+        reduc_ind = list(xrange(1, len(eta.get_shape())))
+        avoid_zero_div = 1e-12
+        if ord == np.inf:
+            eta = tf.clip_by_value(eta, -eps, eps)
+        else:
+            if ord == 1:
+                norm = tf.maximum(avoid_zero_div,
+                                  tf.reduce_sum(tf.abs(eta),
+                                                reduc_ind, keep_dims=True))
+            elif ord == 2:
+                # avoid_zero_div must go inside sqrt to avoid a divide by zero
+                # in the gradient through this operation
+                norm = tf.sqrt(tf.maximum(avoid_zero_div,
+                                          tf.reduce_sum(tf.square(eta),
+                                                        reduc_ind,
+                                                        keep_dims=True)))
+            # We must *clip* to within the norm ball, not *normalize* onto the
+            # surface of the ball
+            factor = tf.minimum(1., eps / norm)
+            eta = eta * factor
+        return eta
+
+
+class FastGradientMethod(Attack):
+    """
+    This attack was originally implemented by Goodfellow et al. (2015) with the
+    infinity norm (and is known as the "Fast Gradient Sign Method"). This
+    implementation extends the attack to other norms, and is therefore called
+    the Fast Gradient Method.
+    Paper link: https://arxiv.org/abs/1412.6572
+    """
+    def __init__(self, model, dtypestr='float32'):
+        """
+        Create a FastGradientMethod instance.
+        Note: the model parameter should be an instance of the
+        cleverhans.model.Model abstraction provided by CleverHans.
+        """
+        if not isinstance(model, Model_Eager):
+            model = CallableModelWrapper(model, 'probs')
+
+        super(FastGradientMethod, self).__init__(model, dtypestr)
+
+    def generate(self, x, **kwargs):
```

Comment:
What is specific to eager in this method? It looks very similar to the one in tf_attacks.py

Reply:
Same as before, lets discuss in the previous comment.